### PR TITLE
Feat: Support model selection for backfilling when running a plan against a dev environment

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

File: docs/reference/cli.md [hunk 1]
```diff
@@ -203,12 +203,13 @@ Options:
...
   --no-auto-categorization  Disable automatic change categorization.
-  --effective-from          The effective date from which to apply forward-only
-                            changes on production.
-  --include-unmodified      Whether to include unmodified models in the target
-                            development environment.
-  --model-selections        Model selection strings to filter the
-                            models that should be included in this plan.
+  --include-unmodified      Include unmodified models in the target
+                            environment.
+  --select-model TEXT       Select specific model changes that should be
```
Comment (tobymao) [inline docs/reference/cli.md:208]: can this support multiple? should it be models?

File: sqlmesh/cli/main.py [hunk 2]
```diff
@@ -287,14 +287,27 @@ def diff(ctx: click.Context, environment: t.Optional[str] = None) -> None:
     multiple=True,
     help="Select specific model changes that should be included in the plan.",
 )
+@click.option(
+    "--backfill-model",
```
Comment (tobymao) [inline sqlmesh/cli/main.py:291]: plurals?

File: sqlmesh/core/plan/definition.py [hunk 3]
```diff
@@ -700,6 +714,13 @@ def _refresh_dag_and_ignored_snapshots(self) -> None:
                 "Revert or apply changes before proceeding with restatements."
             )
 
+        if self._input_backfill_models is not None:
+            if not self.is_dev:
```
Comment (tobymao) [inline sqlmesh/core/plan/definition.py:718]: do we need to be so strict?

File: sqlmesh/cli/main.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/cli/main.py:291]: No, see `--select-model` and `--restate-model`. Multiple entires in CLI look like following:
```
sqlmesh plan --backfill-model model_a --backfill-model "<expression>"
```

File: sqlmesh/core/plan/definition.py (hunk 3 above)
Comment (izeigerman) [reply sqlmesh/core/plan/definition.py:718]: I strongly believe so. Why would anyone want to backfill prod partially?

File: sqlmesh/core/plan/definition.py [hunk 4]
```diff
@@ -700,6 +714,13 @@ def _refresh_dag_and_ignored_snapshots(self) -> None:
...
+                raise PlanError(
+                    "Selecting models to backfill is only supported for development environments."
+                )
+            self._models_to_backfill = set(self.__dag.subdag(*self._input_backfill_models).sorted)
```
Comment (tobymao) [inline sqlmesh/core/plan/definition.py:722]: so this will select all models upstream of models selected to ensure the chain is proper right

Comment (izeigerman) [reply sqlmesh/core/plan/definition.py:722]: this is right

File: sqlmesh/core/plan/definition.py (hunk 3 above)
Comment (tobymao) [reply sqlmesh/core/plan/definition.py:718]: you never know

**Review by tobymao — APPROVED** [review]

File: docs/reference/cli.md [hunk 5]
```diff
@@ -203,12 +203,13 @@ Options:
                             will fail.
   --auto-apply              Automatically apply the new plan after creation.
   --no-auto-categorization  Disable automatic change categorization.
-  --effective-from          The effective date from which to apply forward-only
```
Comment (tobymao) [inline docs/reference/cli.md:206]: why was this removed?

**Comment by tobymao** [comment]
this is a breaking change because you're changing user facing cli

File: docs/reference/cli.md [hunk 6]
```diff
@@ -203,12 +203,13 @@ Options:
...
-                            changes on production.
-  --include-unmodified      Whether to include unmodified models in the target
-                            development environment.
-  --model-selections        Model selection strings to filter the
-                            models that should be included in this plan.
+  --include-unmodified      Include unmodified models in the target
+                            environment.
+  --select-model TEXT       Select specific model changes that should be
+                            included in the plan.
+  --backfill-model TEXT     Backfill only the models whose names match the
```
Comment (tobymao) [inline docs/reference/cli.md:210]: does this support special syntax?

Comment (eakmanrq) [reply docs/reference/cli.md:210]: For example did we decide to not support a syntax for saying directly modified only?

File: sqlmesh/core/plan/definition.py (hunk 3 above)
Comment (eakmanrq) [reply sqlmesh/core/plan/definition.py:718]: I agree with being strict here unless we hear a good use case. The product is opinionated about characteristics that make up prod and we so far this seems like a good bet.

File: docs/reference/cli.md (hunk 6 above)
Comment (izeigerman) [reply docs/reference/cli.md:210]: Depends on what you mean by a special syntax. If you mean `+`, `*`, etc. then yes. If you mean directly modified, then no. I realized supporting directly modified doesn't have much value.

File: docs/reference/cli.md (hunk 5 above)
Comment (izeigerman) [reply docs/reference/cli.md:206]: it was mentioned twice

**Review by eakmanrq — APPROVED** [review]
